src/algorithms/LaurentPoly.jl

```diff
@@ -0,0 +1,212 @@
+###############################################################################
+#
+#   LaurentPoly.jl : Generic algorithms for abstract Laurent polynomials
+#
+###############################################################################
+
+# required methods without default implementation:
+# coeff, setcoeff!, map_coeffs, gen
+
+###############################################################################
+#
+#   Data type and parent object methods
+#
+###############################################################################
+
+base_ring(p::LaurentPolyElem) = base_ring(parent(p))
+
+isdomain_type(::Type{<:LaurentPolyElem{T}}) where {T} = isdomain_type(T)
+
+isexact_type(::Type{<:LaurentPolyElem{T}}) where {T} = isexact_type(T)
+
+function check_parent(a::LaurentPolyElem, b::LaurentPolyElem, throw::Bool = true)
+   c = parent(a) == parent(b)
+   c || !throw || error("incompatible Laurent polynomial rings")
+   return c
+end
+
+
+###############################################################################
+#
+#   Basic manipulation
+#
+###############################################################################
+
+const hashp_seed = UInt === UInt64 ? 0xcdaf0e0b5ade239b : 0x5ade239b
+
+function Base.hash(p::LaurentPolyElem, h::UInt)
+   for i in monomials_degrees(p)
+      c = coeff(p, i)
+      if !iszero(c)
+         h = hash(i, h)
+         h = hash(c, h)
+      end
+   end
+   hash(hashp_seed, h)
+end
+
+# required implementation
+"""
+    monomials_degrees(p::LaurentPolyElem) -> AbstractVector{<:Integer}
+
+> Return a vector containing at least all the degrees of the non-null
+> monomials of `p`.
+"""
+function monomials_degrees end
+
+# return a monomials_degrees vector valid for both polys
+function monomials_degrees(p::LaurentPolyElem, q::LaurentPolyElem)
+   mdp = monomials_degrees(p)
+   mdq = monomials_degrees(q)
+   T = promote_type(eltype(mdp), eltype(mdq))
+   @assert T <: Signed # -one(T) must not wrap around
+   if isempty(mdq)
+      mdp, mdq = mdq, mdp
+   end
+   # if only one is empty, it's now mdp
+   if isempty(mdp) # extrema requires non-empty
+      if isempty(mdq)
+         zero(T):-one(T)
+      else
+         UnitRange{T}(extrema(mdq)...) # mdq not returned for type-stability
+      end
+   else # none is empty
+      minp, maxp = extrema(mdp)
+      minq, maxq = extrema(mdq)
+      min(minp, minq):max(maxp, maxq)
+   end
+end
+
+# like monomials_degrees, but return a "strict" range, whose min/max
+# corresponds to non-zero coeffs, or an empty range for p == 0
+function degrees_range(p::LaurentPolyElem)
+   mds = monomials_degrees(p)
+   T = eltype(mds)
+   minp, maxp = isempty(mds) ? (T(0), T(0)) : extrema(mds)
+   while iszero(coeff(p, minp))
+      minp += 1
+      minp > maxp && return minp:maxp
+   end
+   while iszero(coeff(p, maxp))
+      maxp -= 1
+   end
+   minp:maxp
+end
```

Comment:
I guess this could be specialized for `LaurentPolyWrap` (in a more efficient form)?

Also, how about adding APIs to just query the lead OR the trail degree? Say `lead_degree` and `trail_degree`? Often one needs only one (e.g. in the `lead` method). And perhaps also a `degree` method which would simply call `lead_degree`?

Reply:
> I guess this could be specialized for LaurentPolyWrap (in a more efficient form)?

I generally didn't try to optimize things yet, but I'm not sure here how to significantly speed-up this function for `LaurentPolyWrap` (but it would certainly help if the polynomials were in "canonical" form, i.e. with a maximized `.mindeg` field).

Or indeed, with a `trail_degree` available for the underlying polynomial, a linear scanning of the first non-zero coefficient could be avoided. 
Introducing `lead_degree` and `trail_degree` indeed seems useful. I had not defined `degree` yet as I was not certain this would be an accepted definition.

Reply:
I will introduce `lead_degree` and `trail_degree` as a separate PR as this requires to implement it also for `PolynomialElem`.